lib/src/core/number_to_words/number_to_words.dart

```diff
@@ -1,78 +1,82 @@
+  return numberToWord[number] ?? '';
+}
 
-  words = trim(isNegative ? 'منفی $words' : words);
+/// Gets an [String] input and add negative persian prefix to it
+String _addNegativeSuffix(String number) {
+  return 'منفی $number';
+}
 
-  if (ordinal) words = addOrdinalSuffix(words);
+/// Gets an [int] as input and returns [unitName] for numbers over 1000
+String? _getUnitName(int numberOfZeros) {
+  return numberOfZeros == 0
+      ? ''
+      : numberToWord[int.parse('1${"0" * numberOfZeros}')];
+}
 
-  return words;
+/// Gets an [int] with 3 or less digits as input and converts it to persian
+String? _transformToWord(int number) {
+  if (number == 0) return '';
+  if (number <= 9) {
+    return _getWord(number);
+  } else if (number >= 11 && number <= 19) return _getWord(number);
+
+  var residual = number <= 99 ? number % 10 : number % 100;
+  return residual == 0
+      ? _getWord(number)
+      : '${_getWord(number - residual)} و ${_transformToWord(residual)}';
 }
 
-/// Returns Persian word of the given number in String. Number can contain commas.
-/// You can determine returned string has ordinal suffix or not by `ordinal` flag.
-String? numberToWordsString(String number, {bool ordinal = false}) {
-  if (number.isEmpty) return null;
-  if (number == '0') return zeroFa;
+/// Transform [input] to string inorder to break the number into three digit
+/// parts, get each part's [unitName] and [transformedVal].
+/// Finally combining and returning the result of each part as a [String]
+String? _performer(int number) {
+  if (number <= 999) return _transformToWord(number);
+  var separated = addCommas(number.toString()).split(',');
+  var numberArr = separated
+      .asMap()
+      .entries
+      .map((e) {
+        var transformedVal = _transformToWord(int.parse(e.value, radix: 10));
+        var unitName = _getUnitName((separated.length - (e.key + 1)) * 3);
+        return transformedVal!.isNotEmpty ? '$transformedVal $unitName' : '';
+      })
+      .where((element) => element.length > 1)
+      .toList();
+  return numberArr.join(andFa).trim();
+}
 
-  return _convert(removeCommas(number).toInt(), ordinal);
+/// Get the current integer [input] and pass the absolute int to [_performer]
+/// If [input] is negative, prefix would be added to [_performer] result
+/// If [input] is set to be ordinal, suffix would be added to [_performer] result
+String? _handleResult(int number, bool ordinal) {
+  if (number == 0) return zeroFa;
+  if (number.toString().length > 16) return _numberIsNotValidError();
+  var tmpResult = _isNegative(number)
+      ? _addNegativeSuffix(_performer(number.abs())!)
+      : _performer(number);
+  return ordinal ? addOrdinalSuffix(tmpResult!) : tmpResult;
 }
 
 /// Returns Persian word of the given number in int You can determine
 /// returned string has ordinal suffix or not by `ordinal` flag.
-String numberToWords(int number, {bool ordinal = false}) {
+String? numberToWords(int number, {bool ordinal = false}) {
```

Comment:
Is it possible for this method to return a null value? If not, we can make this null safe, and it would be a great improvement.